### ISSUE TITLE
Tracking issue: Advice Provider refactor (Options for external interface for the VM)

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#409]
### Goal:
Refactor the advice provider as described in the discussion below to improve the external interface to the VM.

```[tasklist]

### Must have

- [x] Advice trait as argument of processor `execute`
- [x] Rename `ProgramInputs` into `AdviceInputs` and move them to the `processor` crate.
- [x] Rename `ProgramOutputs` into `StackOutputs`.
- [x] Add `ProgramInfo` struct and update `verify()` method in the verifier.
- [x] Refactor `write_tape()` in advice provider to use `AdviceSource` as described earlier in this issue.
- [x] Get rid of `MerkleTree`, `MerklePathSet` etc. structs in the `core` create and use the ones from `miden-crypto` instead. (This overlaps with work on the Sparse Merkle Tree in the crypto crate [#36](https://github.com/0xPolygonMiden/crypto/issues/36))
```

### Working group:
@vlopes11, @bobbinth, @tohrnii 

@grjte edit: I've added the above information for reference, but kept @bobbinth's original description in the details below.

### Details

I've started working on implementing a `SYSCALL` operation (for kernel calls) and it led me to think about the overall interface for the VM. One reason for this is that we now need to provide an additional input parameter to specify the kernel against which a program. Another reason is that I think our approach for specifying `AdviceProvider` via `ProgramInputs` struct is probably not flexible enough (i.e., for supporting database-backed advice providers as mentioned in https://github.com/maticnetwork/miden/issues/401#issuecomment-1249529326).

First, for context, a kernel can be represented by a vector of digests (i.e., 4-element words), where each digest is a root of kernel procedure MAST.  So, it could look something like this `Vec<Digest>`.

The verifier needs to be aware of this kernel, and thus, we need to change the `verify()` function to look like this:
```Rust
pub fn verify(
    program_hash: Digest,
    kernel: &[Digest],
    stack_inputs: &[u64],
    outputs: &ProgramOutputs,
    proof: StarkProof,
) -> Result<(), VerificationError> 
```
This looks a bit messy to me - so, we probably should find a way to organize the parameters a bit better. It is probably makes sense to define a separate `Kernel` type. Maybe group inputs and outputs into a single struct. Maybe group kernel and program hash into a single struct etc. Any suggestions on this are welcome.

The prover also needs to be aware of the kernel under which a program is executed. A straightforward approach is to modify the `prove()` function as follows:
```Rust
pub fn prove(
    program: &Program,
    kernel: &Kernel,
    inputs: &ProgramInputs,
    options: &ProofOptions,
) -> Result<(ProgramOutputs, StarkProof), ExecutionError>
```
But as I mentioned above, I think the current approach to building an advice provider from `ProgramInputs` is likely too limiting, and we might actually want to do something like this:
```Rust
pub fn prove(
    program: &Program,
    kernel: &Kernel,
    stack_inputs: &[u64],
    advice_provider: &mut AdviceProvider,
    options: &ProofOptions,
) -> Result<(ProgramOutputs, StarkProof), ExecutionError>
```
This would also mean that we'd need to change the `execute()` function in the processor to look something like this:
```Rust
pub fn execute(
    program: &Program,
    kernel: &Kernel,
    stack_inputs: &[u64],
    advice_provider: &mut AdviceProvider,
) -> Result<ExecutionTrace, ExecutionError> 
```

While the above works, I wonder if it is the best approach. Here are the other possible approaches I thought of.

### Kernel inside Program
Instead of providing kernel as a separate parameter, we could put it inside `Program` struct, which could look something like this:
```Rust
pub struct Program {
    root: CodeBlock,
    kernel: Kernel,
    cb_table: CodeBlockTable,
}
```
This would mean that it would be the assembler who'd need to associate a program with a given kernel. This could be a good thing as we'd be able to verify that all `SYSCALL`'s are valid at assembly time. The overall change could be as simple as adding a new constructor to the assembler - something like:
```Rust
pub fn with_kernel(kernel: Kernel, in_debug_mode: bool) -> Assembler
```
Then any program which would be transformed into MAST using this assembler would be associated with the specified kernel.

This approach would also be logically consistent if we decide to group program hash with kernel for the `verify()` function.

### Stateful VM
Another approach is to replace stateless functions (e.g., `execute()`, `prove()`) with stateful objects. For example, we could define `Processor` struct which could look like this:
```Rust
pub struct Processor {
    kernel: Kernel,
    advice_provider: AdviceProvider,
    in_debug_mode: bool,
}

impl Processor {
    pub fn new(kernel: Kernel, advice_provider: AdviceProvider) -> Self {
        ...
    }

    pub fn execute(&mut self, program: &Program, stack_inputs: &[u64]) -> Result<ExecutionTrace, ExecutionError> {
        ...
    }
}
```

If we do this, we'll also need to do a similar change to the prover. For example, it could look something like this:
```Rust
pub struct MidenProver {
    processor: Processor,
    options: ProofOptions
}

impl MidenProver {
    pub fn new(kernel: Kernel, advice_provider: AdviceProvider, options: ProofOptions) -> Self {
        ...
    }

    pub fn execute(&mut self, program: &Program, stack_inputs: &[u64]) -> Result<(ProgramOutputs, StarkProof), ExecutionError> {
        ...
    }
}
```
Obviously, this is a very significant change to the interface, and while I think it may look a bit cleaner, I wonder if it is worth it.

**Comment by frisitano** [comment]
I tend to favour the stateful VM solution.  It best aligns with my intuition / mental model of how a virtual machine should operate.  In my mind a virtual machine should have some concrete os (kernel) with a storage backend (AdviceProvider) from which I can execute programs.  Had you considered putting all functionality under a single struct? Something like this:

```rust
pub struct MidenVM {
    kernel: Kernel,
    advice_provider: AdviceProvider,
    in_debug_mode: bool,
    options: ProofOptions
}

impl MidenVM {
    pub fn new(kernel: Kernel, advice_provider: AdviceProvider) -> Self {
        ...
    }

    pub fn with_options(self, options: ProofOptions) -> Self {
        ...
    }

    pub fn execute(&self, program: &Program, stack_inputs: &[u64]) -> Result<ExecutionTrace, ExecutionError> {
        ...
    }

    pub fn prove(&self, program: &Program, stack_inputs: &[u64]) -> Result<(ProgramOutputs, StarkProof), ExecutionError> {
        ...
    }

    pub fn verify(&self, program_hash: &Digest, stack_inputs: &[u64], outputs: &ProgramOutputs, proof: StarkProof) -> Result<bool, VerificationError> {
        ...
    }
}
```

Even though the change to the interface is significant, I don't think it will cause too much trouble downstream at this point in time.  Once adoption increases it will become more problematic. 

I think your suggestion of allowing the kernel to be passed to the assembler should be implemented regardless of the model that is chosen.

**Comment by bobbinth** [comment]
I think the above could be the interface for the whole Miden VM, but I'd also like to keep the `Processor` part separate. So, internally it probably makes sense for the `MidenVM` struct to contain `Processor` struct. Or, we could keep the processor stateless, but have the VM stateful.

Agreed regarding passing Kernel to the assembler. I'm actually already working on this.

One other thing, I think method like `execute` and `prove` would need to take `self` as a mutable reference because they may need to modify the advice provider. In regards to this, there might be an argument for the "stateless" approach as it would make it easier to manage the advice provider outside of prove/execute functions. The example I have in mind is as follows:

Let's say we have a database which backs the advice provider. In the stateless model, if execution results in an error we could catch that and rollback a transaction - e.g., similar to this:
```Rust
db.start_transactions();
let (outputs, proof) = match miden::prove(&program, &stack_inputs, &mut db, &options) {
    Ok(result) => {
        db.commit_transaction();
        /* do something */
       result
    }
    Err(err) => {
        db.rollback_transaction();
        /* do something else */
    }
}
```

If we go with the stateful model, these methods might need to be a part of `AdviceProvider` interface, which could look something like this:
```Rust
pub trait AdviceProvider {
    fn read_tape(&mut self) -> Option<Felt>;

    fn write_tape(&mut self, value: Felt);
    fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError>;

    fn get_tree_node(&self, root: Word, depth: Felt, index: Felt) -> Result<Word, ExecutionError>;
    fn get_merkle_path(&self, root: Word, depth: Felt, index: Felt) -> Result<Vec<Word>, ExecutionError>;
    fn update_merkle_leaf(&mut self, root: Word, index: Felt, leaf_value: Word, update_in_copy: bool) -> Result<Vec<Word>, ExecutionError>;

    fn advance_clock(&mut self);

    fn start_transaction() -> Result;
    fn commit_transaction() -> Result;
    fn rollback_transaction() -> Result;
}
```
But then everyone is forced to implement these methods. Which is far from ideal.

**Comment by bobbinth** [comment]
Actually, I'm not sure if the last thing I suggested above is viable - mainly because different databases may have different ways of handling transaction. So, for example, consider transactions in [sled](https://docs.rs/sled/latest/sled/struct.Tree.html#method.transaction). With the stateless approach we could do something like:
```Rust
let (outputs, proof) = db.transaction(|tx_db| {
    let mut adv = SledAdviceProvider::new(tx_db);
    match miden::prove(&program, &stack_inputs, &mut adv, &options) {
        Ok(result) => Ok(result)
        Err(err) => abort(err)?
    } 
})?;
```
But I'm not sure what to do in the stateful approach.

**Comment by frisitano** [comment]
Good catch.  It does look like it will add additional complexity.  Maybe best to avoid such a change at this point.  In that case I personally favour the first approach where dependencies are specified exhaustively without encapsulation of the kernel inside of a `Program` struct.  In my mind the kernel and the program are distinct components and should be represented as such.

Why is the `kernel` represented as a vec of digests as opposed to having a single MAST that encapsulates all kernel procedures?

**Comment by bobbinth** [comment]
> Why is the `kernel` represented as a vec of digests as opposed to having a single MAST that encapsulates all kernel procedures?

Semantically all procedures are independent of each other - so, putting them into a MAST is probably an overkill. We could do a simple Merkle tree of procedures - but that would complicate checking which procedure was called during program execution (which are are planning to do via kernel ROM).

I'm not expecting the number of procedures in a kernel to be too big (e.g., [here](https://github.com/maticnetwork/miden/discussions/352#discussioncomment-3436127) I've identified 9 procedures which would need to be in a transaction kernel. We'll probably need to add more, but I doubt we'll ever go beyond 30 or so) - so, I think keeping them in a list is fine.

**Comment by bobbinth** [comment]
Updating potential trait definition for advice provider based on the latest changes.
```Rust
pub trait AdviceProvider {
    fn read_tape(&mut self) -> Result<Felt, ExecutionError>;
    fn read_tape_w(&mut self) -> Result<Word, ExecutionError>;
    fn read_tape_dw(&mut self) -> Result<[Word; 2], ExecutionError>;

    fn write_tape(&mut self, value: Felt);
    fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError>;

    fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), ExecutionError>;

    fn get_tree_node(&self, root: Word, depth: Felt, index: Felt) -> Result<Word, ExecutionError>;
    fn get_merkle_path(&self, root: Word, depth: Felt, index: Felt) -> Result<Vec<Word>, ExecutionError>;
    fn update_merkle_leaf(&mut self, root: Word, index: Felt, leaf_value: Word, update_in_copy: bool) -> Result<Vec<Word>, ExecutionError>;

    fn advance_clock(&mut self);
}
```

**Comment by bobbinth** [comment]
Another potential interface for the advice provider could be (changing only `write_tape*` methods:
```Rust
pub trait AdviceProvider {
    fn read_tape(&mut self) -> Result<Felt, ExecutionError>;
    fn read_tape_w(&mut self) -> Result<Word, ExecutionError>;
    fn read_tape_dw(&mut self) -> Result<[Word; 2], ExecutionError>;

    fn write_tape(&mut self, source: AdviceSource);

    fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), ExecutionError>;

    fn get_tree_node(&self, root: Word, depth: Felt, index: Felt) -> Result<Word, ExecutionError>;
    fn get_merkle_path(&self, root: Word, depth: Felt, index: Felt) -> Result<Vec<Word>, ExecutionError>;
    fn update_merkle_leaf(&mut self, root: Word, index: Felt, leaf_value: Word, update_in_copy: bool) -> Result<Vec<Word>, ExecutionError>;

    fn advance_clock(&mut self);
}
```

Where `AdviceSource` could be something like this:
```Rust
pub enum AdviceSource {
    Value(Felt),
    AdviceMap(Word),
}
```
This could be more flexible as we add more advice sources in the future (e.g., requesting signatures).

**Comment by vlopes11** [comment]
#640 potentially closes this issue. We have a couple of pending topics that might lead to future improvements.

1) should we allow the CLI endpoints to execute programs with arbitrary advice providers? if yes, we should probably create some interface of communication or a common serialization strategy so we can specify the advice provider as argument of the program call.

2) `AdviceSource` was not introduced in the PR as we can write a word as well as multiple words. it might be added as improvement in future PRs; in that case, an issue must be created.

wdyt?

**Comment by bobbinth** [comment]
I think there is still a few more things here (unless I missed them) needed to make sure we comply with the interface described in https://github.com/0xPolygonMiden/miden-vm/issues/409#issuecomment-1322779186:

1. Rename `ProgramInputs` into `AdviceInputs` and move them to the `processor` crate.
2. Rename `ProgramOutputs` into `StackOutputs`.
3. Add `ProgramInfo` struct and update `verify()` method in the verifier.
4. Refactor `write_tape()` in advice provider to use `AdviceSource` as described earlier in this issue.
5. Get rid of `MerkleTree`, `MerklePathSet` etc. structs in the `core` create and use the ones from `miden-crypto` instead.

The last one should be done after #474 is merged.

**Comment by bobbinth** [comment]
> 1. should we allow the CLI endpoints to execute programs with arbitrary advice providers? if yes, we should probably create some interface of communication or a common serialization strategy so we can specify the advice provider as argument of the program call.

I think it is fine to assume for now that CLI uses only the `MemAdviceProvider`.

**Comment by vlopes11** [comment]
For `5`, https://github.com/0xPolygonMiden/crypto/issues/36 might also be a blocker as we might want to include that as advice set

**Comment by vlopes11** [comment]
@bobbinth what if, instead of `ProgramInfo`, we use `ProgramCommitment`?

**Comment by bobbinth** [comment]
@vlopes11 - eh, I like `ProgramInfo` a bit better mostly because it is more concise.

**Comment by grjte** [comment]
> I think there is still a few more things here (unless I missed them) needed to make sure we comply with the interface described in [#409 (comment)](https://github.com/0xPolygonMiden/miden-vm/issues/409#issuecomment-1322779186):
> 
>     1. Rename `ProgramInputs` into `AdviceInputs` and move them to the `processor` crate.
> 
>     2. Rename `ProgramOutputs` into `StackOutputs`.
> 
>     3. Add `ProgramInfo` struct and update `verify()` method in the verifier.
> 
>     4. Refactor `write_tape()` in advice provider to use `AdviceSource` as described earlier in this issue.
> 
>     5. Get rid of `MerkleTree`, `MerklePathSet` etc. structs in the `core` create and use the ones from `miden-crypto` instead.
> 
> 
> The last one should be done after #474 is merged.

@vlopes11 can you update here with the status of these 5 items when you get a chance?

**Comment by vlopes11** [comment]
@grjte yes, the remainder item is `5`. This will intersect a bit with the work being performed in the tiered SMT https://github.com/0xPolygonMiden/crypto/pull/45 so we might want to leave this item pending and retake it after tiered SMT with its support structs (merkle opening, etc) are finished there.

If we go that way, the only work will be to change the `AdviceSet` enum to wrap the structs from `crypto`.

**Comment by grjte** [comment]
> we might want to leave this item pending and retake it after tiered SMT with its support structs (merkle opening, etc) are finished there.

@vlopes11 this sounds reasonable to me. @bobbinth wdyt?

If we want to do this, can we close this issue and open a new one for this final step?

**Comment by bobbinth** [comment]
I'm fine with creating a separate issue for this - but I do think we should complete it for v0.4 release.

**Comment by bobbinth** [comment]
The last point was closed by #701
